CSharpExtractor/CSharpExtractor/Extractor/Utilities.cs

```diff
@@ -40,7 +47,41 @@ public class Utilities
 			}
 		}
 
-		public static IEnumerable<T> WeakConcat<T>(IEnumerable<T> enumerable1, IEnumerable<T> enumerable2)
+        /// <summary>
+        /// Sample uniform randomly numSamples from an enumerable, using reservoir sampling.
+        /// See https://en.wikipedia.org/wiki/Reservoir_sampling
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="input"></param>
+        /// <param name="numSamples"></param>
+        /// <returns></returns>
+        public static IEnumerable<TSource> ReservoirSample<TSource>(this IEnumerable<TSource> input, int numSamples)
+        {
+            var rng = new Random();
+            var sampledElements = new List<TSource>(numSamples);
+            int seenElementCount = 0;
```

Comment:
Maybe replace this whole logic with a simpler one-liner like:
`inputs.OrderBy(x => Guid.NewGuid()).Take(numSamples);`

?

Reply:
Tried to avoid `OrderBy` for memory/CPU efficiency reasons.

Reply:
OK